### PR TITLE
[WORKAROUND][BOLT][AArch64] Static binary patching for ELF.

### PR DESCRIPTION
**Pull request by paschalis-mpeis**
When patching statically linked binaries, avoid patching GOT entries
that did not belong to the original text section and had an alias.
    

One such special case is the '_init' function that belongs to the '.init'
section. It has '.init' as an alias, which points to the same address of
'_init' in the original binary.
    
This was observed with GNU linker. BOLT normally rejects these cases.

See issue:
- https://github.com/llvm/llvm-project/issues/100096

### CODE REVIEW

**Comment by paschalis-mpeis** [comment]
## Minimal Reproducer:
main.c:
```c
int main() { return 0; }
```

```bash
clang main.c -Wl,-q -static -o out
llvm-bolt out -o out.bolt
./out.bolt
```

**Comment by samolisov** [comment]
As I see, this branch has not been rebased on top of the recently merged changes with the introduction of the `createDummyReturn` function (or the final version of its name).

Please, rebase the branch to distinguish the correct set of changes, what makes the review much simpler.

File: bolt/test/AArch64/patch-elfstatic-libc.test
Comment (samolisov) [inline bolt/test/AArch64/patch-elfstatic-libc.test:1]: Unfortunately, I'm not familiar with the `-static` option and didn't manage to find what exactly it means. Why this option is required here, maybe it applies by default? I see `BC->IsStaticExecutable` in the check you introduced, do they, the option, and the variable's value related? Is `IsStaticExecutable` set to `true` for binaries statically linked with the glibc?

Thank you.

**Review by samolisov — APPROVED** [review]
LGTM, thank you. Also thanks for the test. I personally have a one question related to the test but I'm sure this should not pause the progress.

**Comment by peterwaller-arm** [comment]
Hi, I've filed https://github.com/llvm/llvm-project/issues/100096 to try and capture in detail the issue that this PR is trying to solve.

**Review by yota9 — CHANGES_REQUESTED** [review]
Hello, please rebase the patch for further review

File: bolt/lib/Rewrite/RewriteInstance.cpp [hunk 1]
```diff
@@ -5265,6 +5265,21 @@ void RewriteInstance::patchELFGOT(ELFObjectFile<ELFT> *File) {
...
        ++GOTEntry) {
     if (uint64_t NewAddress = getNewFunctionAddress(*GOTEntry)) {
+      auto *Function = BC->getBinaryFunctionAtAddress(*GOTEntry);
+
+      // On static binaries, avoid patching got entries that did not belong to
+      // the original text section and have an alias. One such special case is
+      // the '_init' function that belongs to the '.init' section. It has
+      // '.init' as an alias, which points to the same address of '_init' in the
+      // original binary.
+      if (BC->IsStaticExecutable && !Function->Aliases.empty() &&
```
Comment (yota9) [inline bolt/lib/Rewrite/RewriteInstance.cpp:5275]: This way the function entry won't be patched either I think? It might also be a problem

**Comment by paschalis-mpeis** [comment]
In the issue below, it was reached a consensus that BOLT [can reject such cases](https://github.com/llvm/llvm-project/issues/100096#issuecomment-2253322315), therefore I am closing this PR.
- https://github.com/llvm/llvm-project/issues/100096


---

I rebased it (as it was a stacked PR) and added some comments in case someone has such a binary and needs a workaround or a starting point.

**Pull request closed without merging**